### PR TITLE
FT: Utapi metrics integration

### PR DESCRIPTION
**Pull request by rahulreddy**
- Integrates utapi server that can be launched using npm run start_utapi
   for listing time series metrics
 - Introduces Utapi client integration with S3 actions

### CODE REVIEW

File: lib/routes/routeHEAD.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import api from '../api/api';
 import routesUtils from './routesUtils';
 
-export default function routeHEAD(request, response, log) {
+export default function routeHEAD(request, response, log, utpai) {
```
Comment (RachedBenMustapha) [inline lib/routes/routeHEAD.js]: utapi would be the correct name :)

File: lib/Config.js [hunk 2]
```diff
@@ -100,6 +100,51 @@ class Config {
             }
         }
 
+        this.utapi = {};
+        if (config.utapi) {
+            if (config.utapi.port) {
```
Comment (LaurenSpiegel) [inline lib/Config.js]: utapi.redis.port?

File: lib/routes/routeDELETE.js [hunk 3]
```diff
@@ -26,6 +40,7 @@ export default function routeDELETE(request, response, log) {
                       return routesUtils.responseNoBody(err, resHeaders,
                         response, null, log);
                   }
+                  _pushMetrics(err, utapi, 'objectDelete', request.bucketName);
```
Comment (alexandre-merle) [inline lib/routes/routeDELETE.js]: does request.bucketName include the object name ?

File: lib/Config.js (hunk 2 above)
Comment (alexandre-merle) [inline lib/Config.js]: here i guess it's utapi port itself :)

File: lib/routes/routeGET.js [hunk 4]
```diff
@@ -11,27 +30,38 @@ export default function routerGET(request, response, log) {
...
+            api.callApiMethod('objectGetACL', request, log, (err, xml) => {
+                _pushMetrics(err, utapi, 'objectGetACL', request.bucketName);
+                return routesUtils.responseXMLBody(err, xml, response, log);
+            });
             // List parts of an open multipart upload
         } else if (request.query.uploadId !== undefined) {
-            api.callApiMethod('listParts', request, log, (err, xml) =>
-                routesUtils.responseXMLBody(err, xml, response, log));
+            api.callApiMethod('listParts', request, log, (err, xml) => {
+                _pushMetrics(err, utapi, 'objectGetACL', request.bucketName);
```
Comment (alexandre-merle) [inline lib/routes/routeGET.js]: listParts :)

File: package.json [hunk 5]
```diff
@@ -33,14 +33,15 @@
     "node-uuid": "^1.4.3",
     "ready-set-stream": "^1.0.0",
     "sproxydclient": "scality/sproxydclient",
+    "utapi": "scality/utapi",
```
Comment (LaurenSpiegel) [inline package.json]: need open...

File: lib/routes/routeDELETE.js (hunk 3 above)
Comment (LaurenSpiegel) [inline lib/routes/routeDELETE.js]: not sending the object name

File: lib/Config.js (hunk 2 above)
Comment (LaurenSpiegel) [inline lib/Config.js]: utapi port needs to be added in config.json

File: lib/routes/routeDELETE.js (hunk 3 above)
Comment (alexandre-merle) [inline lib/routes/routeDELETE.js]: okay :)

Comment (rahulreddy) [inline lib/routes/routeDELETE.js]: Our bucket level metrics does not use object names for the metric data.

File: lib/Config.js (hunk 2 above)
Comment (rahulreddy) [inline lib/Config.js]: Added to config.json

**Comment by rahulreddy** [comment]
@ironman-machine ...

File: package.json [hunk 6]
```diff
@@ -33,14 +33,15 @@
     "node-uuid": "^1.4.3",
     "ready-set-stream": "^1.0.0",
     "sproxydclient": "scality/sproxydclient",
+    "utapi": "scality/utapi#dev/FT/S3Integration",
```
Comment (LaurenSpiegel) [inline package.json]: merge this first

**Comment by rahulreddy** [comment]
@ironman-machine again

**Comment by alexandre-merle** [comment]
:+1: waiting for utapi to be open :)

**Comment by RachedBenMustapha** [comment]
👍

**Comment by rahulreddy** [comment]
Merging this as it's an urgency.

**Pull request merged**